source/developer/webapp-component.rst

```diff
@@ -178,3 +178,129 @@ Testing your Component
+
+.. code-block:: javascript
+  :linenos:
+
+  test('should pass handleSubmit', () => {
+      const newOnSubmit = jest.fn();
+      const newUpdateSection = jest.fn();
+      const props = {...requiredProps, onSubmit: newOnSubmit, updateSection: newUpdateSection};
+      const wrapper = mountWithIntl(<EmailNotificationSetting {...props}/>);
+
+      wrapper.instance().handleSubmit();
+
+      expect(newOnSubmit).not.toBeCalled();
+      expect(newUpdateSection).toHaveBeenCalledTimes(1);
+      expect(newUpdateSection).toBeCalledWith('');
+
+      wrapper.find('#emailNotificationNever').simulate('change');
+      wrapper.instance().handleSubmit();
+
+      expect(newOnSubmit).toBeCalled();
+      expect(newOnSubmit).toHaveBeenCalledTimes(1);
+      expect(newOnSubmit).toBeCalledWith({enableEmail: 'false'});
+
+      expect(savePreference).toHaveBeenCalledTimes(1);
+      expect(savePreference).toBeCalledWith('notifications', 'email_interval', '0');
+
+      ...
+      wrapper.find('textarea').simulate('keydown', {
+          preventDefault: jest.fn(),
+          keyCode: Constants.KeyCodes.ENTER,
+          ctrlKey: false
+      });
+
+      expect(patchChannel).toBeCalledWith('fake-id', {purpose: 'purpose'});
+  });
+
+6. Test component's internal or lifecycle methods by having different sets of props.
+
+.. code-block:: javascript
+  :linenos:
+
+  test('should pass componentWillReceiveProps', () => {
+      const nextProps = {
+          enableEmail: true,
+          emailInterval: 30
+      };
+      const wrapper = mountWithIntl(<EmailNotificationSetting {...requiredProps}/>);
+      wrapper.setProps(nextProps);
+
+      expect(wrapper.state('enableEmail')).toBe(nextProps.enableEmail);
+      expect(wrapper.state('emailInterval')).toBe(nextProps.emailInterval);
+
+      ...
+      const shouldUpdate = wrapper.instance().shouldComponentUpdate({show: true});
+      expect(shouldUpdate).toBe(true);
+  });
+
+7. Provide only simple mockup to imported component or library if necessary.
+
+.. code-block:: javascript
+  :linenos:
+
+  jest.mock('actions/user_actions.jsx', () => ({
+      savePreference: jest.fn()
+  }));
+
+  jest.mock('react-router/es6', () => ({
+      browserHistory: {
+          push: jest.fn()
+      }
+  }));
+
+Finally, initiate the following commands:
+
+1. Run `yarn run test:watch`, select `p` and type filename, and see if the component test passed.
+2. Run `yarn run test:coverage`, and open corresponding html report at `coverage/components` folder to see the percentage covered. Update the test if necessary. Note that it's not required to meet 100% coverage of a component especially if it will take unnecessary or complicated mock up. Uncovered lines, statements, branch or functions will just be recorded so that it will be covered by next test layers like integrations or end-to-end testing. 
```

Comment:
I intentionally put blank on this figure since I don't really have a recommended value. It depends whether the code is testable or not (although we should aim that the codes written should be testable).
Maybe we could start to say 80%+ and then anything below will be on a case to case basis.